api-server/src/server/boot/challenge.js

```diff
@@ -82,6 +90,340 @@ export default async function bootChallenge(app, done) {
+    !!challenge &&
+    [
+      challengeTypes.frontEndProject,
+      challengeTypes.backEndProject,
+      challengeTypes.bonfire,
+      challengeTypes.pythonProject,
+      challengeTypes.codeAllyCert,
+      challengeTypes.multifileCertProject
+    ].includes(body.challengeType) &&
+    validBody;
+
+  if (!isProjectValid) {
+    return res.status(400).json({
+      type: 'error',
+      message: 'Invalid project submission.'
+    });
+  }
+  next();
+}
+
+/**
+ * Handles submissions for all challenges.
+ * @param {Request} req
+ * @param {Response} res
+ * @returns {Observable}
+ */
+export function challengesCompleted(req, res, next) {
+  // Challenges should only include `id` and `challengeType`
+  const { user, body } = req;
+
+  const submittedChallenges = [];
+  for (const challenge of body) {
+    const { id, challengeType } = challenge;
+
+    // NOTE: Consider what will happen if this comes with a batch
+    const completedDate = Date.now();
+
+    const completedChallenge = {
+      challengeType, // TODO: A little birdy told me we do not need to store `challengeType` for non-certs?
+      id,
+      completedDate
+    };
+
+    submittedChallenges.push(completedChallenge);
+  }
+
+  const $push = {
+      progressTimestamps: {
+        $each: []
+      },
+      completedChallenges: {
+        $each: []
+      }
+    },
+    $set = {};
+
+  let points = user.points;
+  for (const chal of submittedChallenges) {
+    const oldIndex = user.completedChallenges.findIndex(c => c.id === chal.id);
+    const alreadyCompleted = oldIndex !== -1;
+    const oldChallenge = alreadyCompleted
+      ? submittedChallenges[oldIndex]
+      : null;
+    if (alreadyCompleted) {
+      $set[`completedChallenges.${oldIndex}`] = {
+        ...chal,
+        completedDate: oldChallenge.completedDate
+      };
+    } else {
+      points++;
+      $push.progressTimestamps.$each = [
+        ...$push.progressTimestamps.$each,
+        Date.now()
+      ];
+      $push.completedChallenges.$each = [
+        ...$push.completedChallenges.$each,
+        chal
+      ];
+    }
+  }
+
+  const updateData = {};
+  if ($push.completedChallenges.$each.length) {
+    updateData.$push = $push;
+  }
+  if (!isEmpty($set)) {
+    updateData.$set = $set;
+  }
+
+  user.points = points;
+  const updatePromise = new Promise((resolve, reject) =>
+    user.updateAttributes(updateData, err => {
+      if (err) {
+        return reject(err);
+      }
+      return resolve();
+    })
+  );
+  return Observable.fromPromise(updatePromise)
+    .map(() => {
+      return res.json({
+        submittedChallenges,
+        points
+      });
+    })
+    .subscribe(() => {}, next);
+}
+
+/**
+ * Handles submissions for all certification projects.
+ * @param {Request} req
+ * @param {Response} res
+ * @returns {Observable}
+ */
+export function projectCompleted(req, res, next) {
+  const { user, body } = req;
+  const { files, completedDate = Date.now(), id } = body;
```

Comment:
```suggestion
  const { files, id } = body;
  const completedDate = Date.now();
```
In contrast to the batches, we don't need to trust the client here.